File: BTree/LeafNode.cpp
#include <iostream>
#include <climits>
#include "LeafNode.h"
#include "InternalNode.h"
#include "QueueAr.h"

using namespace std;


LeafNode::LeafNode(int LSize, InternalNode *p,
  BTreeNode *left, BTreeNode *right) : BTreeNode(LSize, p, left, right)
{
  values = new int[LSize];
}  // LeafNode()

void LeafNode::addToLeft(int value, int last)
{
  leftSibling->insert(values[0]);

  for(int i = 0; i < count - 1; i++)
    values[i] = values[i + 1];

  values[count - 1] = last;
  if(parent)
    parent->resetMinimum(this);
} // LeafNode::ToLeft()

void LeafNode::addToRight(int value, int last)
{
  rightSibling->insert(last);

  if(value == values[0] && parent)
    parent->resetMinimum(this);
} // LeafNode::addToRight()

void LeafNode::addToThis(int value)
{
  int i;

  for(i = count - 1; i >= 0 && values[i] > value; i--)
      values[i + 1] = values[i];

  values[i + 1] = value;
  count++;

  if(value == values[0] && parent)
    parent->resetMinimum(this);
} // LeafNode::addToThis()


void LeafNode::addValue(int value, int &last)
{
  int i;

  if(value > values[count - 1])
    last = value;
  else
  {
    last = values[count - 1];

    for(i = count - 2; i >= 0 && values[i] > value; i--)
      values[i + 1] = values[i];
    // i may end up at -1
    values[i + 1] = value;
  }
} // LeafNode:: addValue()


int LeafNode::getMaximum()const
{
  if(count > 0)  // should always be the case
    return values[count - 1];
  else
    return INT_MAX;
} // getMaximum()


int LeafNode::getMinimum()const
{
  if(count > 0)  // should always be the case
    return values[0];
  else
    return 0;

} // LeafNode::getMinimum()


LeafNode* LeafNode::insert(int value)
{
  int last;

  if(count < leafSize)
  {
    addToThis(value);
    return NULL;
  } // if room for value

  addValue(value, last);

  if(leftSibling && leftSibling->getCount() < leafSize)
  {
    addToLeft(value, last);
    return NULL;
  }
  else // left sibling full or non-existent
    if(rightSibling && rightSibling->getCount() < leafSize)
    {
      addToRight(value, last);
      return NULL;
    }
    else // both siblings full or non-existent
      return split(value, last);
}  // LeafNode::insert()

void LeafNode::print(Queue <BTreeNode*> &queue)
{
  cout << "Leaf: ";
  for (int i = 0; i < count; i++)
    cout << values[i] << ' ';
  cout << endl;
} // LeafNode::print()


LeafNode* LeafNode::remove(int value)
{   // To be written by students
  int pos;

  for(pos = 0; pos < count && values[pos] != value; pos++);

  if(pos < count) // found value
  {
    --count;

    for(int i = pos; i < count; i++)
      values[i] = values[i + 1]; // shift left

    if(count < (leafSize + 1)/ 2)
    {
      if(leftSibling)
        return removeWithLeftSibling();
      else
        if(rightSibling)
          return removeWithRightSibling(pos);
    } // if below minimum

    if(pos == 0  && parent)
      parent->resetMinimum(this);
  }  // if value found

  return NULL;  // unmerged signal
}  // LeafNode::remove()


LeafNode* LeafNode::removeWithLeftSibling()
{
  if(leftSibling->getCount() > (leafSize + 1) / 2)
  {
    insert(leftSibling->getMaximum());
    leftSibling->remove(values[0]);

    if(parent)
      parent->resetMinimum(this);

    return NULL;
  } // if can borrow from left sibling
  else // must merge with left sibling
  {
    for(int i = 0; i < count; i++)
      leftSibling->insert(values[i]);

    leftSibling->setRightSibling(rightSibling);

    if(rightSibling)
      rightSibling->setLeftSibling(leftSibling);

    return this; // merged out of existence sentinel
  } // else must merge
} // removeWithLeftSibling()


LeafNode* LeafNode::removeWithRightSibling(int pos)
{
  if(rightSibling->getCount() > (leafSize + 1) / 2)
  {
    insert(rightSibling->getMinimum());
    rightSibling->remove(values[count - 1]);
    if(pos == 0)
      parent->resetMinimum(this);
    return NULL;
  } // if can borrow from right sibling
  else // must merge with right sibling
  {
    for(int i = 0; i < count; i++)
      rightSibling->insert(values[i]);

    rightSibling->setLeftSibling(leftSibling);

    if(leftSibling)
      leftSibling->setRightSibling(rightSibling);
    return this; // merged out of existence sentinel
  } // else must merge
} // removeWithRightSibling()


LeafNode* LeafNode::split(int value, int last)
{
  LeafNode *ptr = new LeafNode(leafSize, parent, this, rightSibling);


  if(rightSibling)
    rightSibling->setLeftSibling(ptr);

  rightSibling = ptr;

  for(int i = (leafSize + 1) / 2; i < leafSize; i++)
    ptr->values[ptr->count++] = values[i];

  ptr->values[ptr->count++] = last;
  count = (leafSize + 1) / 2;

  if(value == values[0] && parent)
    parent->resetMinimum(this);
  return ptr;
} // LeafNode::split()

File: BTree/leaf_node.py
import sys
from btree import BTree
from internal_node import InternalNode
from b_tree_node import BTreeNode
from leaf_node import LeafNode


class LeafNode:
    def __init__(self, lSize, p = None, l = None, r = None):
            self.internalNode = p
            self.leftTNode = l
            self.rightTNode = r
            self.bTNode = BTreeNode(lSize, p, self.leftTNode, self.rightTNode)
            self.values = []
    
    def addToLeft(self, value, last):
        self.bTNode.left = self.values[0]
        for i in range(0, (self.bTNode).count-1):
            self.values[i] = self.values[i+1]
        self.values[(self.bTNode).count-1] = last;
        if self.bTNode.parent: 
            self.bTNode.parent.resetMinimum(self.bTNode.parent)

    def addToRight(self, value, last):
        self.bTNode.right = last
        if (value == self.values[0] and (self.bTNode.parent)):
            self.bTNode.parent.resetMinimum(self.bTNode.parent)
      
    def addToThis(self, value):
        for i in range( (self.bTNode).count - 1, 0 , -1): 
            if self.values[i] > value:
                self.values[i+1] = self.values[i]
        
        self.values[i+1] = value;
        self.bTNode.count = self.bTNode.count + 1
        if (value == self.values[0] and self.bTNode.parent):
            self.bTNode.parent.resetMinimum(self.bTNode.parent)

    def addValue(self, value, last):
        if value > self.values[(self.bTNode).count-1]:
            last = value;
        else:
            last = self.values[(self.bTNode).count-1]
        for i in range((self.bTNode).count - 2, 0 , -1):
            if self.values[i] <= value:
                break
            self.values[i+1] = self.values[i]
        self.values[i+1] = value

    def getMaximum(self):
        if (self.bTNode).count > 0:
            return self.values[ (self.bTNode).count - 1 ]
        else:
            return sys.maxsize

    def getMinimum(self):
        if (self.bTNode).count > 0:
            return self.values[0]
        else:
            return 0

    def insert(self, value):
        last = 0;
        if ((self.bTNode).count < self.bTNode.leafSize):
            self.addToThis(value)
            return None

        self.addValue(value, last)
        
        if(self.bTNode.left and (self.bTNode.left.count < self.bTNode.leafSize)): #not sure what comes after "and"
            self.addToLeft(value, last)
            return None
        elif (self.bTNode.right and (self.bTNode.right.count < self.bTNode.leafSize)): #not sure what comes after "and"
            self.addToRight(value, last)
            return None
        else:
            return self.split(value, last) #not sure what split is 

    def remove(self, value):
        pos = 0;
        for pos in range(0, self.bTNode.count): #I don't understand this line
            if self.values[pos] == value:
                break
            if (pos < self.bTNode.count):
                self.bTNode.count -= 1
        for i in range(pos, self.bTNode.count):
            self.values[i] = self.values[i+1]
    
        if self.bTNode.count < (((self.bTNode).leafSize + 1) / 2):
            if self.bTNode.left:
                return self.removeWithLeftSibling()
            elif self.bTNode.right:
                return self.removewithRightSibling(pos)
    
        if pos == 0 and self.bTNode.parent:
            self.bTNode.parent = ((self.bTNode).parent).resetMinimum(self.bTNode.parent)
    
        return None


    def removeWithLeftSibling(self):
        if (self.bTNode.left.count > (self.bTNode.leafSize +1)/2):
            self.insert( self.bTNode.left.getMaximum())
            self.bTNode.left.remove(self.values[0]) #not sure 
        if (self.bTNode.parent):
            self.bTNode.parent.resetMinimum(self.bTNode.parent)
            return None
        else:
            for i in range(0, self.bTNode.count):
                self.bTNode.left.insert(self.values[i]) #not sure 
                self.bTNode.left.setRightSibling(self.bTNode.right) # unsure
            if (self.bTNode.right):
                self.bTNode.right.setLeftSibling(self.bTNode.left)
        return self #I think self.bTNode.right is 'this' in this case but I'm not sure

    def removeWithRightSibling(self, pos):
        if(self.bTNode.right.count > ((self.bTNode.leafSize) +1 )/ 2):
            self.insert(self.bTNode.right.getMinimum() ) #this is definitely wrong
            self.bTNode.right.remove(self.values[(self.bTNode).count - 1])
        if pos == 0:
            self.bTNode.parent.resetMinimum(self.bTNode.parent)
            return None
        else:
            for i in range(0, self.bTNode.count):
                self.bTNode.right.insert(self.values[i])
                self.bTNode.right = self.bTNode.setLeftSibling(self.bTNode.left)
            if (self.bTNode.left):
                self.bTNode.left = self.bTNode.setRigthSibling(self.bTNode.right)
            return self.bTNode.left




    def split (self, value, last):
        newPtr =  LeafNode(self.bTNode.leafsize, self.bTNode.parent, self, self.bTNode.right)
        
        if (self.bTNode.right):
            self.bTNode.right = self.bTNode.setLeftSibling(newPtr)
        self.bTNode.right = newPtr
        
        current = newPtr.bTNode.count
        for i in range(  ((((self.bTNode).leafSize) +1)/2), self.bTNode.leafSize):
            newPtr.values[current] = self.values[i] #idkk
            newPtr.bTNode.count += 1
            current = newPtr.bTNode.count
        newPtr.values[current] = last
        newPtr.bTNode.count += 1
        newPtr.bTNode.count = (self.bTNode.leafSize + 1) / 2
        
        if value == self.values[0] and self.bTNode.parent:
            self.bTNode.parent.resetMinimum(self)
        
        return newPtr

    def print_LeafNode(self, address_queue):
        print("Leaf: ")
        for i in range(0, self.bTNode.count):
            print(self.values[i] + ' ')
